packages/valory/skills/decision_maker_abci/behaviours/base.py

```diff
@@ -217,48 +240,106 @@ def check_balance(self) -> WaitableConditionType:
         self.context.logger.info(f"The safe has {native} xDAI and {collateral}.")
         return True
 
+    def send_message(
+        self, msg: Message, dialogue: Dialogue, callback: Callable
+    ) -> None:
+        """Send a message."""
+        self.context.outbox.put_message(message=msg)
+        nonce = dialogue.dialogue_label.dialogue_reference[0]
+        self.shared_state.req_to_callback[nonce] = callback
+        self.shared_state.in_flight_req = True
+
+    def _handle_get_strategy(self, message: IpfsMessage, _: Dialogue) -> None:
+        """Handle get strategy response."""
+        strategy_req = self._inflight_strategy_req
+        if strategy_req is None:
+            self.context.logger.error(f"No strategy request to handle for {message=}.")
+            return
+
+        files = list(message.files.values())
+        n_files = len(files)
+        if n_files != 1:
+            self.context.logger.error(
+                f"Executables for strategies should be in single files. "
+                f"Found {n_files} files for strategy {strategy_req} instead."
+            )
+
+        # store the executable and remove the hash from the mapping because we have downloaded it
+        strategy_exec = files[0]
+        self.shared_state.strategies_executables[strategy_req] = strategy_exec
+        self.shared_state.strategy_to_filehash.pop(strategy_req)
+        self._inflight_strategy_req = None
+
+    def download_next_strategy(self) -> None:
+        """Download the strategies one by one.
+
+        The next strategy in the list is downloaded each time this method is called.
+
+        We download all the strategies,
+        because in the future we will perform some complicated logic,
+        where we utilize more than one, e.g., in case the default fails
+        or is weaker than another depending on the situation.
+
+        :return: None
+        """
+        if self._inflight_strategy_req is not None:
+            # there already is a req in flight
+            return
+        if len(self.shared_state.strategy_to_filehash) == 0:
+            # no strategies pending to be fetched
+            return
+        for strategy, file_hash in self.shared_state.strategy_to_filehash.items():
+            self.context.logger.info(f"Fetching {strategy} strategy...")
+            ipfs_msg, message = self._build_ipfs_get_file_req(file_hash)
```

Comment:
why do this? this class has access to the IPFS helper methods, you can use generators for it and avoid having callbacks?

Reply:
Yeah this is an overkill, not sure why I chose this path. 
I opened #170 as this works for now.